src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -243,27 +363,60 @@ private void Unregister()
             }
 
             Interop.Error errorCode;
-            bool unregistered = !_engine.TryRegister(_fileDescriptor, _registeredEvents, SocketAsyncEvents.None, _handle, out errorCode);
+            bool unregistered = _engine.TryRegister(_fileDescriptor, _registeredEvents, SocketAsyncEvents.None, _handle, out errorCode);
             _registeredEvents = (SocketAsyncEvents)(-1);
 
-            if (!unregistered && errorCode != Interop.Error.EBADF)
+            if (unregistered || errorCode == Interop.Error.EBADF)
```

Comment:
Consider adding a comment about the errorCodes that exist for which you don't want to free the _handle.